lib/openssl/buffering.rb

```diff
@@ -352,22 +346,32 @@ def eof?
 
   def do_write(s)
     @wbuffer = Buffer.new unless defined? @wbuffer
-    @wbuffer << s
-    @wbuffer.force_encoding(Encoding::BINARY)
+    @wbuffer.append_as_bytes(s)
```

Comment:
The `force_encoding(Encoding::BINARY)` in the buffer class was there for a reason, I can only assume. The underlying char pointer is given to the `SSL_write()` function, which has no concept of encoding, but I think it's used elsewhere. It could have to do with the `puts` and `print` methods, which also use `<<` with string arguments, if the arguments have incompatible encodings this will raise an error. It also uses `s.sub!`, if `s` has contents that isn't compatible with its encoding it will raise an error, I think.

Reply:
It was there so that the buffer is always encoded as BINARY, after you call `<<` the encoding may change.

But `append_as_bytes` was added exactly for that use case, it's like `<<` but ensure the encoding is preserved.

Reply:
But now the encoding is not necessarily binary encoding because you removed the `force_encoding` call in the Buffer class's `initialize` method. Now, the buffer can be any encoding and not all calls to `<<` were changed to `append_as_bytes`. Other functions still use this `<<` method and this can result in errors. I think it's better to keep the call to `force_encoding(binary)` in the buffer's `initialize` method and change all `<<` to append_as_bytes in all functions.

Reply:
> But now the encoding is not necessarily binary encoding

It is. `String.new` gives you a binary string:

```ruby
>> String.new.encoding
=> #<Encoding:BINARY (ASCII-8BIT)>
```

> not all calls to << were changed to append_as_bytes

I thought that was the case but I'll have another check, Either way I should probably either make `<<` raise or make it an alias to `append_as_bytes`.

Reply:
> Either way I should probably either make << raise or make it an alias to append_as_bytes

Yeah that makes sense. I think in `puts` there are still calls, in `print` as well and in `fill_rbuff`.

Reply:
I just made these other methods aliases.